### ISSUE TITLE
Index used for OOV 'UNK' token from Tokenizer has a breaking change from 2.1.4 to 2.3.0.

### ISSUE THREAD

**Issue by espears1** [tensorflow/tensorflow#45941]
In Keras 2.1.4, the OOV token added to [Tokenizer](https://faroit.com/keras-docs/2.1.4/preprocessing/text/#tokenizer) will create a new vocabularly index that is 1 unit larger than the largest actual word index. Overall, in 2.1.4, you have index 0 as a reserved token index, index 1 through N as word indexes for actual words in a vocabulary, and index N + 1 as the 'UNK' OOV index.

In Keras 2.3.0, the OOV token uses index 1 instead of N + 1. This affects any pass through use of the tokenizer as well, such as when using `texts_to_sequences`.

This is a breaking change that makes Keras code written for 2.1.4 incompatible with 2.3.0 (for example, tests validating the existence of the 'UNK' token will fail, and re-training some models will leads to wrong word indices in applications that take e.g. an embedding vector learned for the UNK token and use it for other types of processing. Such code will now incorrectly be taking an embedding vector for a real word at index N + 1. Fine-tuning or transfer learning of a model with an Embedding layer that relies on this OOV behavior will also fail, since any new usage of `keras.preprocessing.text.Tokenizer` (such as `texts_to_sequences`) to process inputs into the embedding layer will have swapped the meaning of index N + 1 (previously learned weights for OOV) into index position 1 - and any auxiliary code users have written to extract word embeddings or to do feature importance studies at the level of individual word vectors will be wrong). 

There's currently no API support to always access into the embedding layer via some other mechanism that abstracts away the position of the OOV index, especially since users can provide their own custom token string to use - so unfortunately the actual index position itself is an exposed part of the API of Tokenizer.

I'm curious why this breaking change didn't require updating Keras to 3.0.0 at the time, given that it is breaking the version contracts of semver.

Are there any configuration options to restore the old OOV index behavior (N + 1 instead of 1) when using keras 2.3.0?

Tracking this down through code history, you can see the original version of the preprocessing handling here (using N + 1 as the index):

- https://github.com/keras-team/keras-preprocessing/blame/ea4b8e16d48e2522e6b497d7df9c04aedc63fc7b/keras_preprocessing/text.py#L238

and then you can see the new implementation here, specifically setting the OOV index to 1 instead of N + 1:

- https://github.com/keras-team/keras-preprocessing/commit/878a605a6418c8105ac4c9226945a18ce9e3bf57#diff-45a6438ee723ad37ca89a89ab8d76a4b3e645c3f70d7ec16a5a35fd80a6379c5R230

The way it manifests as a breaking change seems to affect both `keras_preprocessing` as a separate package and transitively `keras` according to the way `keras` sets version pins on `keras_preprocessing`.

**Comment by amahendrakar** [comment]
@espears1,
In order to expedite the trouble-shooting process, could you please provide a minimal code snippet to reproduce the issue reported here. Thanks!

**Comment by espears1** [comment]
@amahendrakar Can you elaborate on what kind of example you'll need? I linked all the source code in my original post which shows the explicit code change creating the bug. Essentially, _any possible_ use of `keras.preprocessing.text.Tokenizer` (particularly `texts_to_sequences`) will have a logically breaking difference between 2.1.4 and 2.3.0 if you examine which position of the Tokenizer's vocabulary houses the OOV token, as described above.

**Comment by rossmessing** [comment]
I'm not sure the TF2.3 label applies - this is about the jump from Keras 2.1.4 to 2.3.0

**Comment by rossmessing** [comment]
Here's a minimal sample:
```
from keras.preprocessing.text import Tokenizer
tokenizer = Tokenizer(50, oov_token="Unknown")
tokenizer.fit_on_texts(["I am a sequence of words", "I am another sequence"])
print(tokenizer.texts_to_sequences(["I am a sequence of characters"]))
```

With Keras 2.2.0 it outputs `[[1, 2, 4, 3, 5, 8]]`
With Keras 2.2.1 it outputs `[[2, 3, 5, 4, 6, 1]]`

**Comment by espears1** [comment]
@rossmessing Thank you - great example!

@jvishnuvardhan @amahendrakar consider the above example and what happens if you use this tokenization scheme to fine-tune a model with an Embedding layer. Suppose the model was originally trained in keras 2.2.0. Within the Embedding layer, index `8` is for out-of-vocabulary words, and the embedding weights for that token index are learned for that purpose. Meanwhile, index `1` weights are learned for `'I'`. Now later if you come back and try to fine-tune in keras 2.2.1, the OOV words will be causing modifications to weights originally learned for `'I'` in index position 1.

This is just one use case. There are many others where you must actually look at the index positions of the output of `texts_to_sequences` and then use them to index into other companion arrays of metadata, pre-trained word vectors, etc.

What it means is that within the output of `texts_to_sequences`, the position of `N + 1` (where `N` is the vocabulary size) has a semantic meaning for the user in 2.2.0 and earlier. It's functionally a part of the API because a user may be required to depend on that specific indexing choice, and this example above shows that the API "breaks" from 2.2.0 to 2.2.1 because of the new treatment of the OOV token's index.

**Comment by Saduf2019** [comment]
@espears1 
Is this still an issue, can you please check in the latest tf and let us know.

**Comment by espears1** [comment]
@Saduf2019 Yep, it's still an issue. It's a backward-compatibility-breaking issue so it affects all TF after the break from 2.2.0 to 2.2.1. Ideally even non-latest published minor versions should receive a new patch bump for it to be fixed, though that might be intractable.

Models trained with 2.2.0 can't have their vocab loaded for fine-tuning in 2.2.1 or onward.

**Comment by sachinprasadhs** [comment]
tf.keras.preprocessing.text.Tokenizer will be deprecated in future version since it does not operate on Tensors, and is most unlikely to get any update.
We recommend you to use tf.keras.layers.TextVectorization which has similar functionality which operate on Tensors.
For details see [here](https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/text/Tokenizer?version=nightly). Thanks!
